services/src/main/java/org/keycloak/userprofile/LegacyUserProfileProvider.java

```diff
@@ -0,0 +1,206 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.userprofile;
+
+import org.jboss.logging.Logger;
+import org.keycloak.common.util.ObjectUtil;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.representations.idm.UserAttributesRepresentation;
+import org.keycloak.services.messages.Messages;
+import org.keycloak.userprofile.util.UpdateUserProfileContext;
+import org.keycloak.userprofile.validation.StaticValidators;
+import org.keycloak.userprofile.validation.UserProfileValidationResult;
+import org.keycloak.userprofile.validation.ValidationChainBuilder;
+import org.keycloak.userprofile.validation.ValidationScenario;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * @author <a href="mailto:markus.till@bosch.io">Markus Till</a>
+ */
+public class LegacyUserProfileProvider extends AbstractUserProfileProvider {
+
+    private static final Logger logger = Logger.getLogger(LegacyUserProfileProvider.class);
+
+
+    public static final String PROVIDER_ID = "legacy-user-profile";
+
+    @Override
+    public String getId() {
+        return PROVIDER_ID;
+    }
+
+    @Override
+    public UserProfileValidationResult validate(UserAttributesRepresentation rep, UpdateUserProfileContext user, ValidationScenario validationScenario) {
+        RealmModel realm = this.session.getContext().getRealm();
+
+        ValidationChainBuilder builder = ValidationChainBuilder.builder();
+        switch (validationScenario) {
+            case UserResource:
+                break;
+            case IdpReview:
+                addBasicValidators(builder, !realm.isRegistrationEmailAsUsername());
+                break;
+            case Account:
+            case RegistrationProfile:
+            case UpdateProfile:
+                addBasicValidators(builder, !realm.isRegistrationEmailAsUsername() && realm.isEditUsernameAllowed());
+                addSessionValidators(builder);
+                break;
+            case RegistrationUserCreation:
+                addUserCreationValidators(builder);
+                break;
+        }
+        return new UserProfileValidationResult(builder.build().validate(rep, user));
+    }
+
+    private void addUserCreationValidators(ValidationChainBuilder builder) {
+        RealmModel realm = this.session.getContext().getRealm();
+
+        if (realm.isRegistrationEmailAsUsername()) {
```

Comment:
This is needed because of the "chained action" approach in the registration. May it should be considered to adapt the registration flow in order to remove this switch so we can go here with configuration only.